src/client.rs

```diff
@@ -0,0 +1,162 @@
+use cosmian_crypto_core::{
+    reexport::rand_core::SeedableRng, Aes256Gcm, CryptoCoreError, CsRng, Dem,
+    FixedSizeCBytes, Instantiable, Nonce, RandomFixedSizeCBytes, SymmetricKey,
+};
+use rand::{Rng, RngCore};
+use std::io::{Error, ErrorKind};
+
+use crate::{btree::DataItem, oram::BUCKET_SIZE};
+
+pub struct ClientORAM {
+    pub stash: Vec<DataItem>,
+    csprng: CsRng,
+    key: SymmetricKey<{ Aes256Gcm::KEY_LENGTH }>,
+}
+
+impl ClientORAM {
+    pub fn new() -> ClientORAM {
+        let mut csprng = CsRng::from_entropy();
+        let key = SymmetricKey::new(&mut csprng);
+
+        ClientORAM {
+            // Empty stash at initialization.
+            stash: Vec::new(),
+            csprng,
+            key,
+        }
+    }
+
+    pub fn gen_key(&mut self) {
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
+
+    pub fn generate_dummies(
+        &mut self,
+        nb_dummies: usize,
+        block_size: usize,
+    ) -> Result<Vec<DataItem>, Error> {
+        if !(nb_dummies + 1).is_power_of_two() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Number of blocks shall be power of 2 minus one".to_string(),
+            ));
+        }
+
+        // Number of leaves is 2^(l-1) if number of blocks is 2^l - 1.
+        let nb_leaves = (nb_dummies + 1) / 2;
+        let mut dummies = Vec::new();
+
+        let cipher = Aes256Gcm::new(&self.key);
+
+        // FIXME - encrypt fixed dummy value instead of encrypting randoms.
+        for _ in 0..nb_dummies * BUCKET_SIZE {
+            // Generate new random dummy data.
+            let mut dummy_data = vec![0; block_size];
+            self.csprng.fill_bytes(&mut dummy_data);
+
+            // Generate new nonce for encryption.
+            let nonce = Nonce::new(&mut self.csprng);
+
+            // Encrypt dummies to provide correct MAC for later decryption.
+            let ciphertext_res =
+                cipher.encrypt(&nonce, &dummy_data, Option::None);
+
+            if ciphertext_res.is_err() {
+                return Err(Error::new(
+                    ErrorKind::Interrupted,
+                    Result::unwrap_err(ciphertext_res).to_string(),
+                ));
+            }
+
+            let encrypted_dummy =
+                [nonce.as_bytes(), ciphertext_res.unwrap().as_slice()].concat();
+
+            // FIXME- uniform generation for now. Is fine but dummies' paths do
+            // not need to be generated at random.
+            let path = self.csprng.gen_range(0..nb_leaves);
+
+            dummies.push(DataItem::new(encrypted_dummy, path as u16));
```

Comment:
You will generate a `u16` if you convert `nb_leaves` to `u16`. But I think it is okay to use a `usize` for the path (I feel like a `u16` may be too small). Plus, as is, you may have an integer overflow during this cast (constrain `nb_dummies` so that it can't happen or add a check on its size).
```suggestion
            let path = self.csprng.gen_range(0..nb_leaves);

            dummies.push(DataItem::new(encrypted_dummy, path));
```

Reply:
I think you are right, I expected no more than 2^16 levels hence u16 for the height, but path being a u16 is wrong  and will overflow (thus panic) when you insert more than 2^17 items.

Reply:
Done in 3be5899d52f3417bdf3412511396b7a280794661